### PR TITLE
feat(contract-verifier): Support Vyper toolchain for EVM bytecodes

### PR DESCRIPTION
**Pull request by slowli**
## What ❔

- Supports the Vyper toolchain for EVM bytecodes in the contract verifier.
- Adds some more unit tests for the verifier (e.g., testing multi-file inputs, Yul contracts, abstract contract errors etc.).

## Why ❔

Part of preparations to support EVM bytecodes throughout the codebase.

## Checklist

- [x] PR title corresponds to the body of PR (we generate changelog entries from PRs).
- [x] Tests for the changes have been added / updated.
- [x] Documentation comments have been added / updated.
- [x] Code has been formatted via `zkstack dev fmt` and `zkstack dev lint`.

### CODE REVIEW

File: core/lib/contract_verifier/src/compilers/mod.rs [hunk 1]
```diff
@@ -31,11 +75,17 @@ fn parse_standard_json_output(
         let errors = errors.as_array().unwrap().clone();
         if errors
             .iter()
-            .any(|err| err["severity"].as_str().unwrap() == "error")
+            .any(|err| err["severity"].as_str() == Some("error"))
         {
+            // FIXME: collected messages contain warnings as well; is this intentional?
```
Comment (slowli) [inline core/lib/contract_verifier/src/compilers/mod.rs:80]: Subj. On the one hand, it sort of makes sense to include warnings from the DevEx perspective, but the fact that they are not distinguished from errors (both are stored as message strings) is somewhat confusing.

File: core/lib/contract_verifier/src/compilers/vyper.rs [hunk 2]
```diff
@@ -0,0 +1,115 @@
...
+        let sources = match req.source_code_data {
+            SourceCodeData::VyperMultiFile(s) => s,
+            other => unreachable!("unexpected `SourceCodeData` variant: {other:?}"),
+        };
+        Ok(Self {
+            contract_name,
+            file_name,
+            sources,
+            optimizer_mode: if req.optimization_used {
+                // FIXME: values have differing semantics compared to other compilers; is this OK?
```
Comment (slowli) [inline core/lib/contract_verifier/src/compilers/vyper.rs:34]: Subj. It probably would make sense to make the optimizer mode dependent on the specified compilers on the type level, or at least check correspondence in the API.

File: core/lib/contract_verifier/src/compilers/mod.rs (hunk 1 above)
Comment (popzxc) [reply core/lib/contract_verifier/src/compilers/mod.rs:80]: I think it's OK for now to not spend extra time on it. I don't think it was intentional, but I also don't think it's a big issue.

File: core/lib/contract_verifier/src/compilers/vyper.rs (hunk 2 above)
Comment (popzxc) [reply core/lib/contract_verifier/src/compilers/vyper.rs:34]: Probably; but the same as with above issue -- it works today, so, it's probably not that big of an issue.

**Review by popzxc — DISMISSED** [review]

**Review by perekopskiy — APPROVED** [review]

**Pull request merged**